README

                             The Modules Package
                                Version 3.1
                                
                               John L. Furlan
                               jlf@behere.com

			        Peter W. Osel
                                 pwo@Osel.DE

                                Jens  Hamisch
                          Jens.Hamisch@Strawberry.COM

                                  R.K. Owen
                              rk@owen.sj.ca.us


                               2  October 1996
                                2  June 2000

1. Introduction
_______________

This directory contains a copy of The Modules Package that uses Tcl (Tool
Command Language) developed by John Ousterhout at Sun Microsystems.  You
must obtain a copy of Tcl (at least version 8.3).  Tcl should be available
either at the same site you obtained The Modules Package or from
http://dev.scriptics.com/ .

For an introduction to The Modules Package, see the paper, "Modules: Providing
a Flexible User Environment", in the Proceeding of the 1991 Large Installation
Systems Administrators USENIX Conference (USENIX LISA V).  A copy of the paper
is included in Postscript form in doc/Modules-Paper.ps.  The paper describes
the concepts behind Modules and a prototype/proof-of-concept implementation
based on shell scripts being sourced into the user environment.  The
implementation details have changed, but the package concepts haven't.


2. Documentation
________________

The ./doc directory contains both the paper and man pages describing the
user's and the module writer's usage.
Look at ChangeLog for detailed information regarding changes.

Big changes to 3.2:
    o  Using the automake/autoconf mechanism for creating the configure script.

    o  Have the configure script test and set more things with regards to
       /bin/sh and /bin/csh characterstics.

    o  Lots of code clean-up.

    o  Made the tests conditional on configure options.

    o  Handle the silent changes in Tcl 8.4

Big changes to 3.1:
    o  Changed the license to GPL

    o  Source code cross-referencing documented in ./xref/*.html files.

Big changes since 3.0 prior to 3.1:
    o  Ported to Linux.

    o  Removed any dependence on low-level Tcl routines.

    o  Can use environment variable references in MODULEPATH for indirection.

    o  Can switch between different versions of the module command.

    o  Fixed "module whatis" to work as advertised.

    o  Added the INSTALL document to aid in first-time and upgrading
       installations.

    o  Fixed "module avail" and "module list" to show levels deeper than 2.

    o  Added "is-loaded" and "module-info specified" to modulefile commands.

Big changes for 3.0:
    o  Testsuite added

    o  Support for Tcl/TclX 7.6 and Tk 4.2

    o  Introduction of user levels for generation of error messages

    o  Enhanced module logging facilities using stdout, stderr, files
       and the syslog

    o  Enhanced module trace feature

    o  Modulefile configuration management. Symbolic module file names
       and aliases

    o  Hierarchical modulecmd rc-file structure

    o  Command line switches

    o  Traceable output for module list and module command (configurable
       and on request)

    o  Active dependecy resolution. This is experimental at the moment!!!

Big changes for 2.3:
    o  Support for Tcl/TclX 7.5 and Tk 4.1

    o  Code cleanup. Added inline documentation. Unification of source
        module layouts.

    o  Added alloc tracer for locating memory leaks

    o  Added 'uname domain'.

    o  Added M_HELP as a modetype, so that 'module-info mode [help]' will
	be possible

    o  Optional support for extended Tcl (tclX) 7.x

    o  Modules now uses the standard Tcl initialization (init.tcl),
	enabling autoloading of Tcl functions.

Big changes for 2.2:
    o  Installation changed to use GNU's configure script (version 2.x).

    o  Updated to support Tcl 7.x

    o  Paths and modulefile names with '+'s in them should work now.

    o  Multi-level paths should work.  Well, at least better than 2.0.

    o  "module load ." doesn't dump core.

    o  Support for "perl" output.

    o  Ability to exit without having modulefile listed as "loaded".
        Use an argument to exit other than 0 or 1.

    o  More stringent checking of user-level command names.

    o  A sysconf.h file to help with multi-platform support.
        This area and the installation process still needs work.

    o  A new and improved cmdXResource.c

    
Big changes for 2.0:

    o  Added a new environment variable, _LOADED_MODULEFILES_, which keeps
        track of which file and directory each modulefile was loaded from.
        This is necessary for locating modulefiles if the MODULEPATH variable
        has changed or if a full pathname was specified as the modulefile to
        load.

    o  Added an update sub-command which attempts to reload all of the
        currently loaded modulefiles.  When Modules is first initialized a
        snapshot of the environment is saved into a file in the user's home
        directory.  When the update sub-command is run, the environment
        variables are restored to their earlier state and the modulefiles are
        reloaded.  Only the variables that modulefiles touch are changed.
        
    o  Added an purge sub-command which unloads all currently loaded
        modulefiles.
    
    o  set-alias should be fixed and should work for all shell types.  In the
        case of the Bourne shell-variants, it creates functions.

    o  If multiple modulefiles are listed to be added or removed and one of
        them has an error, only that modulefile will fail.  All of the other
        modulefiles will load or unload successfully (assuming they don't in
        turn have errors).

    o  INCOMPATABILITY ISSUE:  stdout is not tied to stderr any more.  So, if
        you to a 'puts stdout' in any of your modulefiles, that will go
        directly to the shell.  You should change all puts statements that
        goto stdout to goto stderr.
        
    o  avail prints out all of the modulefiles by recursively looking through
        all directories under every entry in MODULEPATH.
        Only files containing the "#%Module" magic cookie
        header and that don't end in '~' will be listed.  Other files will be
        ignored.

    o  A .version file in a modulefile directory will be parsed to find out
        which modulefile is the default version for the directory name.

    o  Locating modulefiles is recursive and ignores files ending in '~' and
        which don't have the "#%Module" magic header.

    o  The switch command really works now.

    o  Added uname command for fast access to system information.  Not sure if
        uname is portable.

    o  Unload using a directoryname will look for a version that is already
        loaded instead of the default version specified by the .version file
        or by the highest lexicographical name in the directory.

    o  Added ability for modulefile-specific help and modulefile-specific
        display.



3. Building and Installing Modules
__________________________________
(Read the INSTALL file for a more recent description of installation
procedures.)

First, you'll need to decide on a location to keep Modules on your network.

The path name to the initialization files must be the same on all of the
systems using Modules.  This is because your users' .cshrc or other shell
startup files must hard code this directory in order to source their Modules
initialization file.

You will also need to choose a location for the 'modulecmd' program and all of
your modulefiles.  The location of these can be different on different
systems, but it is not recommended.  These paths are used in the Modules
initialization files.  If you decide not to use this structure, then you
will need to edit the initialization files by hand.  Use the --prefix
configure option to control the installation of Modules.  See below for more
configure options.

Example:
	If you decide to keep Modules in /depot/Modules, run
            "configure --prefix=/depot/Modules" 

	initialization file   -->  /depot/Modules/init
	modulefiles           -->  /depot/Modules/modulefiles
	modulecmd             -->  /depot/Modules/bin
 	manual pages          -->  /depot/Modules/man
 	Tcl libraries	      -->  /depot/Modules/lib/tcl	(if any)

The Modules Package is written in conformant ANSI C and should build on any
UNIX machine with an ANSI C compiler like the GNU C compiler.  Check the
MACHINES file for a list of machines and OSs Modules has been test built on.
The module error logger requires the ANSI C 'stdarg.h' being installed. At
the moment there's no support for traditional vargargs handling using 
'varargs.h'.

To build Modules, first build Tcl and run the tests provided with Tcl to
verify it built correctly.

If you want to use the testsuite, you should install dejagnu-1.2 or better
before configuring modules. Be sure to have your PATH pointing to 'runtest'
when calling 'configure'.

Then run the GNU autoconf script, ./configure.
A Makefile will be automatically generated.

Some Modules-specific configure options...  (See also ./configure --help)

Autoconf @variable@'s can be passed along to the configure script, in
particular for setting paths.  The one of interest is @VERSION@ for
using the current module version. e.g.  --prefix=/usr/local/Modules/@VERSION@

------------------------------------------------------------------------
    --with-etc-path=<path>      use etc path=<path>		[/etc]
    --with-skel-path=<path>     use skel path=<path>		[/etc/skel]

  o Place to find the csh.modules * profile.modules files and where
    the new user dot files are (both are needed by add.modules script).
  o The skeleton user dot files need to point to the etc path for
    setting up modules environment on login.

------------------------------------------------------------------------
    --with-split-size=<size>	With split size=<size>		[TEST]
    --without-split-size	Without  - " -

  o Most C Shell implementations have a size requirement on the length of
      the line which can be evaluated.  If you run into very long
      environment changes, you may run into this problem.  Select size
      to be something less than the size your C Shell will accept.
      The configure script will test this in your /bin/csh upto
      10,000 characters.  If you do not specify anything configure
      will set the split-size for you if the limit it finds is less
      than 4000 characters.  This can be overridden by specifying
      --with-split-size='some_value' specifying --without-split-size
      will disable this feature.  Just specifying --with-split-size
      without a value will default to a value of 1000.

------------------------------------------------------------------------
    --enable-shell-funcs	With sh functions		[TEST]
    --disable-shell-funcs	Without  - " -

  o Does your Bourne Shell support functions?  If so, then the modulefile
      set-alias will work properly.  If not, aliases will not be available
      when using the Bourne Shell.
      The default is to test your /bin/sh when configuring build, but
      specifying either option will override the tested characteristic.

------------------------------------------------------------------------
    --enable-shell-alias	With sh functions		[TEST]
    --disable-shell-alias	Without  - " -

  o Does your Bourne Shell support aliases?  A true Bourne Shell doesn't
      but most are just a link to some other Bourne-like shell.  If so,
      then modules will use the alias mechanism when using set-alias.
      If not, aliases will not be available unless sh functions are
      supported when using the Bourne Shell.
      The default is to test your /bin/sh when configuring build, but
      specifying either option will override the tested characteristic.

------------------------------------------------------------------------
    --enable-shell-eval		With shell alias eval		[DEF]
    --disable-shell-eval	Without  - " -

  o This specifies whether aliases will be evaluated by the shell or
      sourced from a temporary file.  Some shells can't handle the
      specification of aliases in the eval line very well.  To get around
      this problem, when aliases are set by a modulefile, a temporary file
      is created and then sourced into the shell.  This guarentees the
      aliases are evaluated correctly, but may leave lots of cruft
      in the /tmp directory, which is generally not a good idea.
 
------------------------------------------------------------------------
    --enable-free		With free() calls
    --disable-free		Without  -" -		[DEF]

  o Since the modulecmd is transient, disabling calls to free() can provide
      some small performance enhancement since the memory usage is low and
      the program will exit rapidly.  If you want the calls to free() back
      in for some reason, set this option.

------------------------------------------------------------------------
     --with-static		Link static (don't use dynamic libraries)
     --without-static		Use dynamic libraries		[DEF]

  o A note.  I've found that at least the X11 libraries should be linked
      statically.  If you encounter problems, e.g. getting some warnings
      about version mismatch of dynamic libraries, link the modulecmd
      statically.

  o If you're using Tcl/TclX 7.5 on an SunOS or Solaris box, the dynamic
      loader library (libdl.so.1) will be liked in. Since this is not
      available as a staic library, using  --without-static is recommended.

------------------------------------------------------------------------
     --with-tclx		With extended TCL (tclX) commands
     --without-tclx		Without  - " -			[DEF]

  o Extended Tcl is a superset of standard Tcl and is built alongside
      the standard Tcl sources.  It adds many new commands to standard
      Tcl, e.g. Unix Access Commands, File I/O Commands, String and
      Character Manipulation Commands, ...  See the manual pages that
      come with tclX for a complete list of added features.

------------------------------------------------------------------------
     --with-module-path=<path>	With user defined MODULEPATH
     --without-module-path	Default $prefix/modulefiles	[DEF]

  o Set the MODULEFILE environment variable in all initialization
      files to the given <path> (colon separated!).  The default
      is to use $prefix/modulefiles (which usually contains version
      specific modulefiles).
  o Every site should use this to specify where local modulefiles
      are placed which are independent of module versions.

------------------------------------------------------------------------
    --with-version-path         use module path=<path>
                                [/usr/local/Modules/versions]

  o Place to put module version modulefiles for switching between different
      versions of the module command.
------------------------------------------------------------------------
     --with-autoload-path=<path>	directories where Tcl will search
					for libraries to load
     --without-autoload-path		Default $prefix/lib/tcl	[DEF]

  o <path> will be prepended to Tcl's auto_path variable.  These
      directories are searched for library functions that can be
      autoloaded.  Note that this path has to be *space* separated and
      enclosed in double quotes!!
      The default is to use "$prefix/lib/tcl".

------------------------------------------------------------------------
     --with-tcl=<path>		directory containting Tcl configuration
				(tclConfig.sh)
     --with-tclx=<path>		directory containting TclX configuration
				(tclxConfig.sh)

  o the tclConfig.sh file was created when Tcl was built, and has pointers
  to where the libraries and header files were installed. It is usually
  found in $prefix/lib, where $prefix is the directory you installed Tcl.
  (Same for TclX too.)

------------------------------------------------------------------------
     --with-debug=<level>		Set up the debug level to the
					specified value
     --without-debug			Disable debugging [DEF]

  o The debug level will be set to the specified value. Available values
      may be looked up in 'modules_def.h'. Debugging may be configured in
      order to trace:

	- module commands
	- TCL interpreter initialization
	- callback functions
	- module file localization
	- utility functions

      Enable this feature for debugging only! The higher the debugging level,
      the greater the amount of runtime information printed! All debug mes-
      sages will be spooled to <stderr>.

      The debug-level cannot be changed at runtime at the moment! It is
      configured as a fixed value at compile time!

------------------------------------------------------------------------
     --enable-version-magic             set .version magic requirement[DEF]
     --disable-version-magic            unset requirement

  o Enable or disable the requirement that .version files need the
      "#%MOdule1.0" magic cookie at the file start to be properly recognized
      and parsed.

------------------------------------------------------------------------
NO LONGER AVAILABLE
     --enable-parseable-msgs		Configure parseable error messages
     --disable-parseable-msgs		Use the traditional ones

  o Switch between traditional error messages in multi-line format or parse-
      able ones that look like the GNU messages. The parseable message format
      is:

	module(line):Message-Type:Message-Number: Text

      with Message-Type being one of:

	INFO, WARNING, PROBLEM, ERROR, FATAL, PANIC, DEBUG

      and an unique message id in decimal format.

------------------------------------------------------------------------
      --enable-logging                  Enable the module logging feature
      --disable-logging                 No module logging [DEF]
      --with-log-facility-verbose[=<facility>] 
      --with-log-facility-info[=<facility>]
      --with-log-facility-debug[=<facility>]
      --with-log-facility-debug[=<facility>]
      --with-log-facility-warn[=<facility>]
      --with-log-facility-problem[=<facility>]
      --with-log-facility-problem[=<facility>]
      --with-log-facility-fatal[=<facility>]

NEWS
* Note: Implied Consent -
  If you send in a code patch or fix of one type or another you consent
  to have your name and email address listed here unless you explicitly
  note otherwise.  Whether your name & email is actually listed depends
  on how extensive the change.  (If it's a minor code, document change,
  or spelling correction your name will probably not be listed.)

****************************   Release 3.2 ******************************
Modules 3.2.13
Apr  7, 2020				Xavier Delaruelle (xavier.delaruelle@cea.fr)
    * Fix buffer overflow when working with large MODULEPATH (thanks to
      Dominique Martinet for his help) (fix issue#311)
    * Install: dynamically set gitlog2changelog.py script shebang
    * Fix unset-alias logic (Martin Schuster) (fix issue#321)

Modules 3.2.12
Mar 23, 2019				Xavier Delaruelle (xavier.delaruelle@cea.fr)
    * Fix some C-string issues (Tim Haines) (fix issue#221)
    * Fix configure.ac not to search for a TclX installation if
      --without-tclx is set (fix issue#217)

Modules 3.2.11
Oct 16, 2017				Xavier Delaruelle (xavier.delaruelle@cea.fr)
	* Fixed configure.ac bug with AM_INIT_AUTOMAKE (Alexey Brodkin)
	* Fix compilation with -Werror=implicit-function-declaration
	  (Orion Poplawski)
	* Fix default version set in nested directory (Jan Synacek)
	* Make Python init script Python 3 compatible (Slavek Kabrda)
	* Fix unload from loaded modulefile (Filip Krska)
	* Add support for Tcl 8.6 (Orion Poplawski)
	* Fix build with -Werror=format-security (Orion Poplawski)
	* Fix minor compiler warnings (Alastair McKinstry)
	* Add completion to avail command on bash (Orion Poplawski)
	* Fix "module clear" command (Orion Poplawski)
	* Comment out stray "module use" in example modules file when not
	  using versioning (Orion Poplawski)
	* Fix modulecmd path in init files (Orion Poplawski)

Modules 3.2.10
Dec 21, 2012				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the module switch with custom delimiters (Tyson Whitehead)
	* If a bash shell check if interactive or not for alias/functions
	* Fix the flags with regards to recursive load/unload
	* Fix perl.pm init file typo (bug 3497092)
	* Moved from cvs to git (and now: modulefile/module-git)
	* Fix the module swap with the --delim option (Tyson Whitehead)
	* Increased the "update" buffer (Leon Kos)
	* Use the locale for sorting modulefiles by the avail command
	* Fixed the segfault problem with Tcl_RegExpCompile() (Kenneth Hoste,
	  Tyson Whitehead)
	* Updated the FSF address in LICENSE.GPL (Jan Synacek)
	* Forked the Tcl-only version to be maintained by Kent Mein.
	* Added the "module-info type" query to return 'C' to indicate that
	  the modulecmd executable is the "C" version.
	  (The Tcl-only version will return 'Tcl'.)
	* Search /usr/lib64 for tclConfig.sh also (Jan Synacek)
	* Tested loading 20 other modulefiles from a single one
	* Have ChangeLog generated by a script due to Marcus D. Hanwell

Modules 3.2.9
Nov 24, 2011				R.K. Owen (rk@owen.sj.ca.us)
	* Fix the coexistence of load/remove flags (Martin Siegert)
	* Fixed an overlapping string copy (Wilson Snyder & Orion Poplawski)
	* Isolated calls to the new interpreter and get & set env.vars.
	* Crucial bug fix of same by (Orion Poplawski).
	* Optionally use Tcl memory check routines for malloc/realloc.
	* Fixed the "module purge" memory corruption (Poor Yorick)
	* Added support for Ruby (Tammo Tjarks)

Modules 3.2.8
Oct  1, 2010				R.K. Owen (rk@owen.sj.ca.us)
	* Handle the .modulerc/.version in deep modulefile dirs,
	  Patches thanks to Lindsay Todd, and others.
	* Note that .version takes precedence over .modulerc files.
	* Renamed init/python to init/python.py and updated with
	  anonymous patch to eliminate 'sed' references
	* Renamed init/perl to init/perl.pm
	* Added brief section to module.1 on invoking modules for perl & python
	* Added Nathaniel Waisbrot's patch for CMake support with tests.
	* Fixed the MANPATH in modulefiles/modules to use configure generated
	  values.
	* Added the configure --with-man-path so a system specific MANPATH
	  could be used.  Some "man"s radically change their search path
	  when a MANPATH exists and when not.
	  The default is "/usr/man:/usr/share/man" (when both exists).
	* Purging a purged module list does not result in an error.
	* Fixed the configure script for searching for tclConfig.sh
	  and handle --without-x in the tests.

Modules 3.2.7
Jul 30, 2009				R.K. Owen (rk@owen.sj.ca.us)
	* Have init/bash export the module function (thanks to Roy Dragseth)
	* Fixed swap bug when prepending or appending multiple times
	  (fix thanks to Gavin Walker), added tests for same.
	* Have purge remove loaded modules in reverse order
	  (patch contributed by carriees)
	* Added a batch of patches from Gerrit Renker
	  - Goes through each of the two manpages, bringing them up to date.
	  - Fix an error in configure.ac caused by misspelling
	  - Fix typos in help messages and source-code comments
	  - Fix the setup of $MODULESPATH to not have a trailing ':'
	  - Fix segmentation violation occurring when the shelltype is absent
	  - Fix a condition where 'make uninstall' does not cleanse a directory
	  - Fix a build problem which set the executable bit on 'data' files
	  - Cleans up noise in '.modulespath' file
	  - Replaces external sed filter script with internal Makefile.am rule
	  - Adds missing 'zsh' case in etc/global
	  - simplify (t)csh initialisation
	  - Add bash auto-completion for the 'module' command
	  - Adds new 'chdir' modulefile command (sets cwd on module load)
	* versions a & b fixed some bugs

Modules 3.2.6
Oct 29, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the locate_module.c:realloc() calls which didn't properly
	  take the element size into account (Fix thanks to Ron Isaacson).
Modules 3.2.5
Feb 14, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Wrapped all malloc() calls to make sure the size is never 0.
	  This fixes a problem spotted by Francesco Nigro under AIX,
	  and diagnosed by G Lehnert.

cmdInfo.c
/*****
 ** ** Module Header ******************************************************* **
 ** 									     **
 **   Modules Revision 3.0						     **
 **   Providing a flexible user environment				     **
 ** 									     **
 **   File:		cmdInfo.c					     **
 **   First Edition:	1991/10/23					     **
 ** 									     **
 **   Authors:	John Furlan, jlf@behere.com				     **
 **		Jens Hamisch, jens@Strawberry.COM			     **
 **		R.K. Owen, <rk@owen.sj.ca.us> or <rkowen@nersc.gov>	     **
 ** 									     **
 **   Description:	The Tcl module-info routine which provides informa-  **
 **			tion about the state of Modules as the modulefile    **
 **			is being parsed. 				     **
 ** 									     **
 **   Exports:		cmdModuleInfo					     **
 ** 									     **
 **   Notes:								     **
 ** 									     **
 ** ************************************************************************ **
 ****/

/** ** Copyright *********************************************************** **
 ** 									     **
 ** Copyright 1991-1994 by John L. Furlan.                      	     **
 ** see LICENSE.GPL, which must be provided, for details		     **
 ** 									     ** 
 ** ************************************************************************ **/

static char Id[] = "@(#)$Id$";
static void *UseId[] = { &UseId, Id };

/** ************************************************************************ **/
/** 				      HEADERS				     **/
/** ************************************************************************ **/

#include "modules_def.h"

/** ************************************************************************ **/
/** 				  LOCAL DATATYPES			     **/
/** ************************************************************************ **/

/** not applicable **/

/** ************************************************************************ **/
/** 				     CONSTANTS				     **/
/** ************************************************************************ **/

/** not applicable **/

/** ************************************************************************ **/
/**				      MACROS				     **/
/** ************************************************************************ **/

/** not applicable **/

/** ************************************************************************ **/
/** 				    LOCAL DATA				     **/
/** ************************************************************************ **/

static	char	module_name[] = __FILE__;

/** ************************************************************************ **/
/**				    PROTOTYPES				     **/
/** ************************************************************************ **/

/** not applicable **/


/*++++
 ** ** Function-Header ***************************************************** **
 ** 									     **
 **   Function:		cmdModuleInfo					     **
 ** 									     **
 **   Description:	Callback function for 'module-info'		     **
 ** 									     **
 **   First Edition:	1991/10/23					     **
 ** 									     **
 **   Parameters:	ClientData	 client_data			     **
 **			Tcl_Interp	*interp		According Tcl interp.**
 **			int		 objc		Number of arguments  **
 **			Tcl_Obj		*objv[]		Argument array	     **
 ** 									     **
 **   Result:		int	TCL_OK		Successful completion	     **
 **				TCL_ERROR	Any error		     **
 ** 									     **
 **   Attached Globals:	g_flags			These are set up accordingly **
 **						before this function is	     **
 **						called in order to control   **
 **						everything.		     **
 **			g_specified_module	The module name from the     **
 **						command line.		     **
 **			g_current_module	The module which is handled  **
 **						by the current command	     **
 ** 									     **
 ** ************************************************************************ **
 ++++*/

int cmdModuleInfo(
	ClientData client_data,
	Tcl_Interp * interp,
	int objc,
	Tcl_Obj * CONST84 objv[]
) {
	char           *arg0, *arg1, *arg2;
	char           *s, *t, buf[BUFSIZ];

    /**
     **  Parameter check
     **/
	arg0 = Tcl_GetString(objv[0]);
	if (objc < 2) {
		if (OK != ErrorLogger(ERR_USAGE, LOC, arg0, "info-descriptor ",
				      "descriptor-args", NULL))
			return (TCL_ERROR);	/** ---- EXIT (FAILURE) ---> **/
	}
	arg1 = Tcl_GetString(objv[1]);
	if (objc > 2)
		arg2 = Tcl_GetString(objv[2]);
	if (!strcmp(arg1, "type")) {
    /**
     **  'module-info type'
     **   returns 'C' to refer to the C-version of modules
     **   (the Tcl-only version returns 'Tcl')
     **/
		Tcl_SetResult(interp, "C", TCL_STATIC);

	} else if (!strcmp(arg1, "flags")) {
    /**
     **  'module-info flags'
     **  returns the current flag value
     **/
		char            tmpbuf[6];
		sprintf(tmpbuf, "%d", g_flags);
		Tcl_SetResult(interp, tmpbuf, TCL_VOLATILE);

	} else if (!strcmp(arg1, "mode")) {
    /**
     **  'module-info mode'
     **  without suggestion this will return the name of the state we're in.
     **/
		if (objc < 3) {
			if (g_flags & M_SWSTATE1)
				Tcl_SetResult(interp, "switch1", TCL_STATIC);
			else if (g_flags & M_SWSTATE2)
				Tcl_SetResult(interp, "switch2", TCL_STATIC);
			else if (g_flags & M_SWSTATE3)
				Tcl_SetResult(interp, "switch3", TCL_STATIC);
			else if (g_flags & M_LOAD)
				Tcl_SetResult(interp, "load", TCL_STATIC);
			else if (g_flags & M_REMOVE)
				Tcl_SetResult(interp, "remove", TCL_STATIC);
			else if (g_flags & M_DISPLAY)
				Tcl_SetResult(interp, "display", TCL_STATIC);
			else if (g_flags & M_HELP)
				Tcl_SetResult(interp, "help", TCL_STATIC);
			else if (g_flags & M_WHATIS)
				Tcl_SetResult(interp, "whatis", TCL_STATIC);
			else if (g_flags & M_NONPERSIST)
				Tcl_SetResult(interp, "nonpersist", TCL_STATIC);
			else
				Tcl_SetResult(interp, "unknown", TCL_STATIC);
		} else {
	    /**
	     **  'module-info mode <my_suggestion>'
	     **  results in a boolean return indicating whether this is the
	     **  correct state ...
	     **/
			Tcl_SetResult(interp, "0", TCL_STATIC);

			if (!strcmp(arg2, "load")) {
				if (g_flags & M_LOAD)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "remove")) {
				if (g_flags & M_REMOVE)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "display")) {
				if (g_flags & M_DISPLAY)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "nonpersist")) {
				if (g_flags & M_NONPERSIST)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "help")) {
				if (g_flags & M_HELP)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "whatis")) {
				if (g_flags & M_WHATIS)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "switch1")) {
				if (g_flags & M_SWSTATE1)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "switch2")) {
				if (g_flags & M_SWSTATE2)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "switch3")) {
				if (g_flags & M_SWSTATE3)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "switch")) {
				if (g_flags & M_SWITCH)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else {
				ErrorLogger(ERR_USAGE, LOC, arg0, "mode ",
					    "[load|remove|display|help|whatis|switch1|switch2|switch3|switch|nonpersist]",
					    NULL);
				return (TCL_ERROR); /** -- EXIT (FAILURE) -> **/
			}
		}

	} else if (!strcmp(arg1, "user")) {
    /**
     **  'module-info user'
     **  without suggestion this will return the current user level
     **/
		if (objc < 3) {
			if (UL_NOVICE == sw_userlvl)
				Tcl_SetResult(interp, "novice", TCL_STATIC);
			else if (UL_ADVANCED == sw_userlvl)
				Tcl_SetResult(interp, "advanced", TCL_STATIC);
			else if (UL_EXPERT == sw_userlvl)
				Tcl_SetResult(interp, "expert", TCL_STATIC);
			else
				Tcl_SetResult(interp, "unknown", TCL_STATIC);
		} else {
	    /**
	     **  'module-info user <my_suggestion>'
	     **  results in a boolean return indicating whether this is the
	     **  correct user level ...
	     **/
			Tcl_SetResult(interp, "0", TCL_STATIC);

			if (!strcmp(arg2, "novice")) {
				if (UL_NOVICE == sw_userlvl)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "advanced")) {
				if (UL_ADVANCED == sw_userlvl)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else if (!strcmp(arg2, "expert")) {
				if (UL_EXPERT == sw_userlvl)
					Tcl_SetResult(interp, "1", TCL_STATIC);
			} else {
				ErrorLogger(ERR_USAGE, LOC, arg0, "user ",
					    "[novice|advanced|expert]", NULL);
				return (TCL_ERROR); /** -- EXIT (FAILURE) -> **/
			}
		}

	} else if (!strcmp(arg1, "name")) {
    /**
     **  'module-info name'
     **  returns the name of the current module
     **/
		Tcl_SetResult(interp, g_current_module, TCL_VOLATILE);

	} else if (!strcmp(arg1, "shell")) {
    /**
     **  'module-info shell'
     **  returns the name of the current user shell
     **/
		if (objc < 3) {
			Tcl_SetResult(interp, shell_name, TCL_VOLATILE);
		} else {
			if (!strcmp(arg2, shell_name))
				Tcl_SetResult(interp, "1", TCL_STATIC);
			else
				Tcl_SetResult(interp, "0", TCL_STATIC);
		}

	} else if (!strcmp(arg1, "shelltype")) {
    /**
     **  'module-info shelltype'
     **  returns the name of the current user shelltype family
     **/
		if (objc < 3) {
			Tcl_SetResult(interp, shell_derelict, TCL_VOLATILE);
		} else {
			if (!strcmp(arg2, shell_derelict))
				Tcl_SetResult(interp, "1", TCL_STATIC);
			else
				Tcl_SetResult(interp, "0", TCL_STATIC);
		}

	} else if (!strncmp(arg1, "trace", 5)) {
	/**
	 ** 'module-info trace' -- no longer used so return *undef
	 **/
		/* trace or tracepat */
		Tcl_SetResult(interp, "*undef*", TCL_STATIC);

	} else if (!strcmp(arg1, "alias")) {
    /**
     **  'module-info alias'
     **  returns the value of the passed alias
     **/
		if (objc < 3) {
			if (OK != ErrorLogger(ERR_USAGE, LOC, arg0, "alias ",
					      "name", NULL))
				return (TCL_ERROR); /** -- EXIT (FAILURE) -> **/
		}

		if (AliasLookup((char *)arg2, &s, &t)) {
			stringer(buf, BUFSIZ, s, psep, t, NULL);
			Tcl_SetResult(interp, buf, TCL_VOLATILE);
		} else {
			Tcl_SetResult(interp, "*undef*", TCL_STATIC);
		}

	} else if (!strcmp(arg1, "symbols")) {
    /**
     **  'module-info symbols'
     **  List all symbolic names of the passed or current module file
     **/
		char           *name;

		name = (objc < 3) ? g_current_module : (char *)arg2;

		if (!(s = ExpandVersions(name)))
			Tcl_SetResult(interp, "*undef*", TCL_STATIC);
		else
			Tcl_SetResult(interp, (char *)s, TCL_VOLATILE);

	} else if (!strcmp(arg1, "version")) {
    /**
     **  'module-info version'
     **  Returns the full qualified module name and version of the passed
     **  symbolic version specifier
     **/
		if (VersionLookup((char *)arg2, &s, &t)) {
			if (t) {
				stringer(buf, BUFSIZ, s, psep, t, NULL);
			} else {
				stringer(buf, BUFSIZ, s, NULL);
			}
			Tcl_SetResult(interp, buf, TCL_VOLATILE);
		} else {
			Tcl_SetResult(interp, "*undef*", TCL_STATIC);
		}

	} else if (!strcmp(arg1, "specified")) {
    /**
     **  'module-info specified'
     **   gives the module name as specified on the command line
     **/
		if (g_specified_module) {
			/* TCL_STATIC because it comes from the command line */
			Tcl_SetResult(interp, g_specified_module, TCL_STATIC);
		} else {
			Tcl_SetResult(interp, "*undef*", TCL_STATIC);
		}

	} else {
    /**
     **  unknown command ....
     **/
		if (OK != ErrorLogger(ERR_INFO_DESCR, LOC, arg1, NULL))
			return (TCL_ERROR); /** ------ EXIT (FAILURE) -----> **/
	}

	return (TCL_OK);

} /** End of 'cmdModuleInfo' **/
